### PR TITLE
feat(instrumentation-hapi): support v21

### PR DESCRIPTION
**Pull request by nlochschmidt**
## Which problem is this PR solving?

- https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1383

## Short description of the changes

Continuation of stale #1491

- Updates Hapi dependency to v21
- Uses types shipped with library instead of `@types/hapi__hapi`
- Adds Hapi v21 to the list of tested versions
- Supports tracing when route.options is a function

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-hapi/src/utils.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-hapi/src/utils.ts:131]: I think I'd rather have this written out as nested if-else statements. On first glance, it's somewhat hard to see what's going on here.

**Review by pichlermarc — COMMENTED** [review]
I need to try this out still, but looks good at first glance - thanks for taking care of this :slightly_smiling_face: 
Two comments for now:

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json:53]: This TypeScript project is setup to do lib checks. The alternative is to add `skipLibCheck: true` to `tsconfig.json`

https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1491#issuecomment-1967943017

File: plugins/node/opentelemetry-instrumentation-hapi/src/utils.ts
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/src/utils.ts:131]: I took this from the original PR. I also think that should be nicer to read.

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json]: Judging by the rest of the code base, adding `skipLibCheck: true` seems like the right thing to do.

Change is pushed.

**Comment by nlochschmidt** [comment]
@pichlermarc did you have a chance to look at this yet?

**Comment by nlochschmidt** [comment]
@trentm given your experience with hapi, maybe you could review this PR? 🙏

File: plugins/node/opentelemetry-instrumentation-hapi/src/instrumentation.ts
Comment (trentm) [inline plugins/node/opentelemetry-instrumentation-hapi/src/instrumentation.ts:278]: Do you understand why this statement was added in https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1595 originally? I would *guess* it was left over debugging code, but I don't know.

**Review by trentm — COMMENTED** [review]
LGTM, just the .tav.yml change that I think should be added, and a Q about dropping the `@types/hapi__hapi` dep.  Tests *do* pass, so I don't think my Q about the types needs to hold this up.

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json:64]: These dependencies should have been `devDependencies` in first place, but yes users of a certain typescript version should bring the types that match their version of hapi.

File: plugins/node/opentelemetry-instrumentation-hapi/src/instrumentation.ts
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/src/instrumentation.ts:278]: That's what I though. Since the tests are passing I assumed it was just a left-over. If not I would expect a comment here explaining why this is needed.

**Comment by pichlermarc** [comment]
@nlochschmidt please resolve the conflicts if you have some time. :slightly_smiling_face: 
Doing so will make the workflow run and that'll help allow us to see if the tests are all passing. :slightly_smiling_face:

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json:53]: While this worked for the hapi plugin itself, it causes the `auto-instrumentations-node` build to fail, presumably because that is now doing a transitive lib check. I'll bring back the joi dev-dependency, but I wonder if there are any other ideas on how to solve this?

**Comment by nlochschmidt** [comment]
@pichlermarc good to know that tests only run when there are no conflicts 💡 

I had to bring back the `joi` dev dependency to make the tests pass (see [this comment](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1985#discussion_r1537223412))

**Review by trentm — APPROVED** [review]
LGTM

I defer to @pichlermarc or others if skipLibCheck:true on auto-instrumentations-node might be a preferable change.

**Comment by nlochschmidt** [comment]
> I defer to @pichlermarc or others if skipLibCheck:true on auto-instrumentations-node might be a preferable change.

My 2c: I think adding `joi` as a devDependency is reasonable here, because joi comes [from the hapi project](https://github.com/hapijs/joi). I consider this similar to other plugins (e.g. koa) bringing in multiple dev dependencies from the project it's instrumenting.

Disabling lib check on auto-instrumentation-node seems like a bigger discussion, and I wonder if it's worth waiting for a decision before merging this PR?

**Comment by trentm** [comment]
Just reproducing it myself, the build failure:

```
$ npm run compile
...
       > @opentelemetry/auto-instrumentations-node@0.43.0 compile
       > tsc -p .

       ../../node_modules/@hapi/hapi/lib/types/route.d.ts:2:68 - error TS2307: Cannot find module 'joi' or its corresponding type declarations.

       2 import { ObjectSchema, ValidationOptions, SchemaMap, Schema } from 'joi';
                                                                            ~~~~~

       ../../node_modules/@hapi/hapi/lib/types/server/server.d.ts:4:22 - error TS2307: Cannot find module 'joi' or its corresponding type declarations.

       4 import { Root } from 'joi';
                              ~~~~~


       Found 2 errors.

       npm ERR! Lifecycle script `compile` failed with error:
       npm ERR! Error: command failed
       npm ERR!   in workspace: @opentelemetry/auto-instrumentations-node@0.43.0
       npm ERR!   at location: /Users/trentm/tm/opentelemetry-js-contrib4/metapackages/auto-instrumentations-node
...
```

Note: we already have skipLibCheck on a subset of packages:

```
% rg skipLibCheck
plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json:    "skipLibCheck": true
plugins/node/opentelemetry-instrumentation-mongodb/tsconfig.json:    "skipLibCheck": true
plugins/node/instrumentation-lru-memoizer/tsconfig.json:    "skipLibCheck": true
plugins/node/instrumentation-mongoose/tsconfig.json:    "skipLibCheck": true
plugins/node/instrumentation-cucumber/tsconfig.json:    "skipLibCheck": true
metapackages/auto-instrumentations-web/tsconfig.json:    "skipLibCheck": true
metapackages/auto-instrumentations-web/tsconfig.esm.json:    "skipLibCheck": true,
metapackages/auto-instrumentations-web/tsconfig.esnext.json:    "skipLibCheck": true,
plugins/web/opentelemetry-instrumentation-user-interaction/tsconfig.json:    "skipLibCheck": true
plugins/web/opentelemetry-instrumentation-user-interaction/tsconfig.esm.json:    "skipLibCheck": true,
plugins/web/opentelemetry-instrumentation-user-interaction/tsconfig.esnext.json:    "skipLibCheck": true,
```


My 2c ;), I am fine with either:

1. taking this PR as is, and separately considering using `skipLibCheck:true` in auto-instrumentations-node; or
2. adding skipLibCheck to auto-instrumentations-node now. (I think we should eventually.)

@pichlermarc Did you have a particular opinion?

**Comment by nlochschmidt** [comment]
I would love to see this PR getting merged. @trentm @pichlermarc can I do anything to make that happen?

**Comment by toddtarsi** [comment]
@nlochschmidt - Thanks for carrying this forward! I second that, if there's anything I could do to help here as well, I would be glad to.

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (pichlermarc) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json:64]: Yes I think you're right, they should've been `devDependencies`. Most likely they were part of the public API at some point but when they were removed we forgot to remove them from `dependencies`.

I'm good with removing :+1:

**Comment by pichlermarc** [comment]
> My 2c ;), I am fine with either:
> 
>     1. taking this PR as is, and separately considering using `skipLibCheck:true` in auto-instrumentations-node; or
> 
>     2. adding skipLibCheck to auto-instrumentations-node now. (I think we should eventually.)
> 
> 
> @pichlermarc Did you have a particular opinion?

Sorry for taking so long to come back to this PR. :fearful: 

In my initial review I was just *very confused* that `joi` is part of the public `@hapi/hapi` types, but not listed as a `dependency` or `peerDependency` anywhere in the `@hapi/hapi` package. As a result importing `@hapi/hapi` with TS seems to break by default. (I've never encountered that before as a deliberate choice)

My guess is this choice was maybe made as it can be used without `joi` in plain JS (?)

Anyway, I'm fine to 
- keep the `joi` package as a `devDependency` for compilation
- this means we can remove `skipLibCheck: true` again (locally this compiles fine with `joi` installed)  

I'd prefer a `joi` `devDependency` over having `skipLibCheck: true` added anywhere. IMO we should only add it if there's no other way.

File: plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json:6]:
```suggestion
    "outDir": "build"
```
should still compile IIUC :slightly_smiling_face:

Comment (trentm) [reply plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json:6]: Which means we can remove all changes to this file. The only other changes were indentation.

**Comment by trentm** [comment]
Thanks, Mark.

@nlochschmidt Are you able to drop the change to "plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json" and deal with merge conflicts now? (Let me know if you want a hand with that.)

File: plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/tsconfig.json]: I reverted all changes to this file.

**Comment by nlochschmidt** [comment]
@trentm @pichlermarc PR is up-to-date again.

> _Note: Sorry about the force-push. I know you don't like that. It was just an accidentally pushed commit to `HEAD` that I immediately replaced with the correct one. I hope that's alright 🤞_

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-hapi/package.json:67]:
```suggestion
}

```

**Review by pichlermarc — APPROVED** [review]
Thank you for working on this :+1:

File: plugins/node/opentelemetry-instrumentation-hapi/package.json
Comment (nlochschmidt) [reply plugins/node/opentelemetry-instrumentation-hapi/package.json]: Odd. My editor is setup to use `.editorconfig`, but this wasn't auto-fixed.

**Pull request merged**
